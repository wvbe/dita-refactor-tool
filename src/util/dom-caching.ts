import { evaluateUpdatingExpression, executePendingUpdateList } from 'fontoxpath';
import { Node, serializeToWellFormedString } from 'slimdom';
import { sync } from 'slimdom-sax-parser';

export type FileCacheOptions = {
	resolve: (filePath: string) => string;
	fetch: (filePath: string) => Promise<string>;
	push: (filePath: string, contents: string) => Promise<void>;
};

export class FileCache {
	private stringByName: { [filePath: string]: string | null } = {};
	private domByFileName: { [filePath: string]: Document } = {};
	private options: FileCacheOptions;

	constructor(options: FileCacheOptions) {
		this.options = options;
	}

	public knowsAbout(filePath: string): boolean {
		return !!this.stringByName[filePath];
	}

	public async getString(filePath: string): Promise<string> {
		if (!this.stringByName[filePath] && this.domByFileName[filePath]) {
			// This cache was removed after getDocument returned it (after which it may have been changed by reference).
			this.injectString(
				filePath,
				serializeToWellFormedString((this.domByFileName[filePath] as unknown) as Node)
			);
		}

		if (!this.stringByName[filePath]) {
			// Either this thing was never fetched before, or someone botched busting the cache
			const resolvedFilePath = this.options.resolve(filePath);
			this.injectString(filePath, await this.options.fetch(resolvedFilePath));
		}

		return this.stringByName[filePath] as string;
	}

	public async getDocument(filePath: string): Promise<Document> {
		// NOTE caching on the filePath key, not a resolved file path. This is probably fine for a while.
		if (this.domByFileName[filePath]) {
			return this.domByFileName[filePath];
		}

		const fetchedContent = await this.getString(filePath);

		this.domByFileName[filePath] = (sync(fetchedContent) as unknown) as Document;

		// The DOM may be changed after it is returned, so unset that cache and let getString serialize the DOm back
		// to a string next thime it is asked.
		this.stringByName[filePath] = null;

		return this.domByFileName[filePath];
	}

	public injectString(filePath: string, contents: string) {
		this.stringByName[filePath] = contents;
	}
	public async updateDocument(
		filePath: string,
		updateQuery: string,
		variables: Record<string, any>
	) {
		const { pendingUpdateList } = await evaluateUpdatingExpression(
			updateQuery,
			await this.getDocument(filePath),
			null,
			variables || {},
			{
				debug: true
			}
		);

		let executed = false;
		return {
			pendingUpdateList,
			execute: () => {
				if (executed) {
					throw new Error(`Cannot execute pending update list twice`);
				}
				executed = true;
				return executePendingUpdateList(pendingUpdateList);
			}
		};
	}

	public bustFile(filePath: string): void {
		delete this.stringByName[filePath];
		delete this.domByFileName[filePath];
	}

	public async writeFile(filePath: string, contents?: Document): Promise<void> {
		const dom = contents || (await this.getDocument(filePath));
		return this.options.push(filePath, serializeToWellFormedString((dom as unknown) as Node));
	}
}
