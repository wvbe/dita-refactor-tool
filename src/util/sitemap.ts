import {
	evaluateXPath,
	evaluateXPathToBoolean,
	evaluateXPathToNodes,
	evaluateXPathToString,
	evaluateXPathToStrings
} from 'fontoxpath';
import { FileCache } from './dom-caching';

/**
 * Represents a flat sitemap node, aka. does not have parent/child relationships. Does contain all the information
 * needed to give it a name or URL, and find its contents.
 */
export type SitemapNodeType = {
	id: string;
	navtitle: string;

	target?: string;
	resource: boolean;
};

/**
 * Hierarchical sitemap nodes.
 */
export type SitemapTreeType = SitemapNodeType & {
	children: SitemapTreeType[];
};

type MinifiedSitemapTree = number | MinifiedSitemapTree[];

// Exported for testing
export async function getUniqueMapsInTree(
	fileCache: FileCache,
	rootFilePath: string,
	forceRefresh: boolean
): Promise<string[]> {
	const knownFilePaths: string[] = [];
	const queuedFilePaths: string[] = [rootFilePath];

	let filePath;
	while ((filePath = queuedFilePaths.shift())) {
		knownFilePaths.push(filePath);
		if (forceRefresh) {
			fileCache.bustFile(filePath);
		}
		const dom = await fileCache.getDocument(filePath);
		const newFilePaths = evaluateXPathToStrings('//mapref[@href]/@href', dom).filter(
			nestedFilePath =>
				!knownFilePaths.includes(nestedFilePath) &&
				!queuedFilePaths.includes(nestedFilePath)
		);
		queuedFilePaths.splice(0, 0, ...newFilePaths);
	}

	return knownFilePaths;
}

// Exported for testing
export async function getUniqueItemsFromMap(
	fileCache: FileCache,
	mapFilePath: string,
	forceRefresh: boolean
): Promise<SitemapNodeType[]> {
	if (forceRefresh) {
		fileCache.bustFile(mapFilePath);
	}
	return evaluateXPath(
		`array {
			//(topicref|topichead)/map {
				"id": string(@id),
				"navtitle": string(./topicmeta/navtitle),
				"target": if (@href)
					then string(@href)
					else (),
				"resource": boolean(@processing-role = 'resource-only')
			}
		}`,
		await fileCache.getDocument(mapFilePath)
	);
}

// Exported for testing
export async function getUniqueItemsFromMaps(
	fileCache: FileCache,
	mapFilePaths: string[],
	forceRefresh: boolean
): Promise<SitemapNodeType[]> {
	const results: SitemapNodeType[] = [];
	for await (const filePath of mapFilePaths) {
		results.splice(
			results.length,
			0,
			...(await getUniqueItemsFromMap(fileCache, filePath, forceRefresh))
		);
	}
	return results;
}

// Exported for testing
export async function getItemTree<T>(
	fileCache: FileCache,
	rootFilePath: string,
	boxItem: (node: Node, children: T[]) => T[]
): Promise<T[]> {
	return (async function recurse(node: Document | Node): Promise<T[]> {
		if (node.nodeType === 9) {
			// If the node is the documentNode, go to the documentElement instead
			return recurse((node as Document).documentElement);
		}
		if (evaluateXPathToBoolean('self::mapref', node)) {
			return recurse(await fileCache.getDocument(evaluateXPathToString('@href', node)));
		}

		const childNodes = evaluateXPathToNodes('./(topicref|topichead|mapref)', node);
		const childItems = (
			await Promise.all(childNodes.map(node => recurse(node as Node)))
		).reduce((flat: T[], item: T[]) => flat.concat(...item), []);

		return boxItem(node, childItems);
	})(await fileCache.getDocument(rootFilePath));
}

// Exported for testing
export async function getExpandedItemTree(
	fileCache: FileCache,
	items: SitemapNodeType[],
	rootFilePath: string
) {
	return getItemTree<SitemapTreeType>(fileCache, rootFilePath, (node, children) => {
		if (!evaluateXPathToBoolean('self::topichead or self::topicref', node)) {
			// Some levels are invisible, such as DITAMAPs
			return children;
		}

		const id = evaluateXPathToString('@id', node) || null;
		if (!id) {
			throw new Error(`Sitemap node "${id}" was not indexed before building hierarchy`);
		}

		const item = items.find(item => item.id === id);
		if (!item) {
			return children;
		}

		return [
			{
				...item,
				children
			}
		];
	});
}

// Exported for testing
// @TODO probably remove this
export async function getCompressedItemTree(
	fileCache: FileCache,
	items: SitemapNodeType[],
	rootFilePath: string
) {
	return await getItemTree<MinifiedSitemapTree>(fileCache, rootFilePath, (node, children) => {
		if (!evaluateXPathToBoolean('self::topichead or self::topicref', node)) {
			// Some levels are invisible, such as DITAMAPs
			return children;
		}

		const id = evaluateXPathToString('@id', node) || null;
		if (!id) {
			throw new Error(`Sitemap node "${id}" was not indexed before building hierarchy`);
		}

		const index = items.findIndex(item => item.id === id);
		if (index < 0) {
			return children;
		}

		return [children.length > 0 ? [index, children] : index];
	});
}

// Exported for testing
export function extractItemTree(
	compressedItemTree: MinifiedSitemapTree,
	items: SitemapNodeType[]
): SitemapNodeType[] {
	if (!Array.isArray(compressedItemTree)) {
		compressedItemTree = [compressedItemTree];
	}
	const root = { children: [] };

	(function recurse(
		parent: {
			children: SitemapNodeType[];
		},
		level: MinifiedSitemapTree[]
	) {
		for (let i = 0; i < level.length; ++i) {
			const item = {
				...items[level[i] as number],
				children: []
			};

			if (Array.isArray(level[i + 1])) {
				recurse(item, level[++i] as MinifiedSitemapTree[]);
			}

			parent.children.push(item);
		}
	})(root, compressedItemTree);

	return root.children;
}

export class Sitemap {
	fileCache: FileCache;
	rootFilePath: string;
	mapFilePaths?: string[];
	items?: SitemapNodeType[];

	constructor(fileCache: FileCache, rootFilePath: string) {
		this.fileCache = fileCache;
		this.rootFilePath = rootFilePath;
	}

	async getMaps(forceRefresh?: boolean) {
		if (!this.mapFilePaths) {
			this.mapFilePaths = await getUniqueMapsInTree(
				this.fileCache,
				this.rootFilePath,
				!!forceRefresh
			);
		}
		return this.mapFilePaths;
	}

	/**
	 * Return a flat list of all nodes in the sitemap. Nodes do not have information wrt. their parent or children.
	 */
	public async getNodes(forceRefresh?: boolean) {
		if (!this.items) {
			this.items = await getUniqueItemsFromMaps(
				this.fileCache,
				await this.getMaps(forceRefresh),
				!!forceRefresh
			);
		}
		return this.items;
	}

	/**
	 * Get a hierarchical tree, where all nodes are organized according to their parent/children relationships.
	 */
	public async getTree() {
		return getExpandedItemTree(this.fileCache, await this.getNodes(), this.rootFilePath);
	}
}
