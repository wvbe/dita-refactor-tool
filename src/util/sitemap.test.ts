import { beforeAll, describe, expect, it } from '@jest/globals';
import fs from 'fs-extra';
import path from 'path';
import { FileCache } from './dom-caching';
import { extractItemTree, getCompressedItemTree, Sitemap } from './sitemap';

describe('In sitemap.test.1.xml', () => {
	const fileCache = new FileCache({
		resolve: p => path.join(__dirname, '__data__', p),
		fetch: p => fs.readFile(p, 'utf8'),
		push: async () => {}
	});
	const sitemap = new Sitemap(fileCache, 'sitemap.test.1.xml');

	let tree: any, extracted: any;

	beforeAll(async () => {
		tree = await getCompressedItemTree(
			fileCache,
			await sitemap.getNodes(),
			'sitemap.test.1.xml'
		);
		extracted = extractItemTree(tree, await sitemap.getNodes());
	});

	it('[Sitemap#getMaps] finds the expected amount of maps', async () => {
		expect(await sitemap.getMaps()).toHaveLength(2);
	});

	it('[Sitemap#getTopics] finds the expected amount of items', async () => {
		expect(await sitemap.getNodes()).toHaveLength(9);
	});

	it('[getCompressedItemTree] returns the expected array', async () => {
		expect(tree).toEqual([
			0, //root
			[
				1, // 1
				[
					2, // 1.1
					3, // 1.2
					[
						4 // 1.2.1
					]
				],
				5, // 2
				[
					6, // 2.1
					7, // 2.2
					8 // 2.3
				]
			]
		]);
	});

	it('[Sitemap#getExpandedItemTree] returns the expected array', async () => {
		expect(await sitemap.getTree()).toMatchSnapshot();
	});

	it('[extractItemTree] extracts to the expected object', async () => {
		expect(extracted).toEqual(await sitemap.getTree());
	});
});
