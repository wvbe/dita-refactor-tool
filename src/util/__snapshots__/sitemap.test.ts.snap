// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`In sitemap.test.1.xml [Sitemap#getExpandedItemTree] returns the expected array 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": "main-1.1",
            "navtitle": "Main 1.1",
            "resource": false,
            "target": "file2.xml",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "id": "main-1.2.1",
                "navtitle": "Main 1.2.1",
                "resource": false,
                "target": "file2.xml",
              },
            ],
            "id": "main-1.2",
            "navtitle": "Main 1.2",
            "resource": false,
            "target": "file2.xml",
          },
        ],
        "id": "main-1",
        "navtitle": "Main 1",
        "resource": false,
        "target": "file1.xml",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": "main-2.1",
            "navtitle": "Main 2.1",
            "resource": false,
            "target": "file2.xml",
          },
          Object {
            "children": Array [],
            "id": "nested-1",
            "navtitle": "Nested 1",
            "resource": false,
            "target": "file4.xml",
          },
          Object {
            "children": Array [],
            "id": "nested-2",
            "navtitle": "Nested 2",
            "resource": false,
            "target": "file4.xml",
          },
        ],
        "id": "main-2",
        "navtitle": "Main 2",
        "resource": false,
        "target": "file3.xmlxml",
      },
    ],
    "id": "main-root",
    "navtitle": "Main root",
    "resource": false,
    "target": "file4.xml",
  },
]
`;
